File: twidere/src/main/java/org/mariotaku/twidere/adapter/ArrayRecyclerAdapter.java
package org.mariotaku.twidere.adapter;

import android.support.v7.widget.RecyclerView.Adapter;
import android.support.v7.widget.RecyclerView.ViewHolder;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/**
 * Created by mariotaku on 14/10/27.
 */
public abstract class ArrayRecyclerAdapter<T, H extends ViewHolder> extends Adapter<H> {

    protected final ArrayList<T> mData = new ArrayList<>();

    @Override
    public final void onBindViewHolder(H holder, int position) {
        onBindViewHolder(holder, position, getItem(position));
    }

    public abstract void onBindViewHolder(H holder, int position, T item);


    public void add(final T item) {
        if (item == null) return;
        mData.add(item);
        notifyDataSetChanged();
    }

    public void addAll(final Collection<? extends T> collection) {
        mData.addAll(collection);
        notifyDataSetChanged();
    }

    public void clear() {
        mData.clear();
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    public T getItem(final int position) {
        return mData.get(position);
    }

    public boolean remove(final int position) {
        final boolean ret = mData.remove(position) != null;
        notifyDataSetChanged();
        return ret;
    }

    public void removeAll(final List<T> collection) {
        mData.removeAll(collection);
        notifyDataSetChanged();
    }

    public void sort(final Comparator<? super T> comparator) {
        Collections.sort(mData, comparator);
        notifyDataSetChanged();
    }
}

File: twidere/src/main/kotlin/org/mariotaku/twidere/adapter/ArrayRecyclerAdapter.kt
package org.mariotaku.twidere.adapter

import android.content.Context
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.bumptech.glide.RequestManager
import java.util.*

/**
 * Created by mariotaku on 14/10/27.
 */
abstract class ArrayRecyclerAdapter<T, H : ViewHolder>(
        context: Context,
        requestManager: RequestManager
) : BaseRecyclerViewAdapter<H>(context, requestManager) {

    protected val data = ArrayList<T>()

    override fun onBindViewHolder(holder: H, position: Int) {
        onBindViewHolder(holder, position, getItem(position))
    }

    abstract fun onBindViewHolder(holder: H, position: Int, item: T)


    fun add(item: T?) {
        if (item == null) return
        data.add(item)
        notifyDataSetChanged()
    }

    fun addAll(collection: Collection<T>) {
        data.addAll(collection)
        notifyDataSetChanged()
    }

    fun clear() {
        data.clear()
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return data.size
    }

    fun getItem(position: Int): T {
        return data[position]
    }

    fun remove(position: Int): Boolean {
        data.removeAt(position)
        notifyItemRemoved(position)
        return true
    }

    fun removeAll(collection: List<T>) {
        data.removeAll(collection)
        notifyDataSetChanged()
    }

    fun sort(comparator: Comparator<in T>) {
        Collections.sort(data, comparator)
        notifyDataSetChanged()
    }
}
